CenterForOpenScience/osf.io
# [ENG-896] EGAP Registration

## Purpose

Create a management command to create the EGAP Registrations from the EGAP draft registrations and projects created by other PRs/JIRA tickets.

## Changes

* Adds `external_registered_date` to Registration model
  * Surfaces `external_registered_date`, if available, in API as `date_created` for a Registration
* Modifies `EmailApprovableSanction` to add a `should_suppress_emails` property based upon the existence of `external_registered_date` with the associated Registration
  * Modifies `ask` and `_on_complete` to suppress emails if `should_suppress_emails` property is set to `True`
    * Moves `ask` from `TokenApprovableSanction` to `EmailApprovableSanction` which allows `ask` to have the email suppression check and fits better within the scope of the classes
      * Change calls to `TokenApprovableSanction.ask` to `EmailApprovableSanction.ask`
* Creates management command to register EGAP DraftRegistrations
  * Registers the DraftRegistrations
  * Adds an `external_registered_date` representing the date the new registration was created in EGAP
  * Create RegistrationApproval or Embargo for each new EGAP Registration
    * Emails requesting approval to admins and notifying non-admins of sanction are suppressed
  * Promote authors on project associated with each new EGAP Registration to Admin

## QA Notes

There is a model change to the `Registration` model which requires a low-risk migration. 

The registration detail and list endpoint should be tested to ensure that if a Registration has a `external_registered_date`, it is provided as the `date_created` field in the API response (/v2/registrations/ or /v2/registrations/<registration id>/).

The suppression of emails to admins and non-admins for Registrations with a valid `external_registered_date` (and lack of suppression for Registrations without an `external_registered_date`) should also be tested

## Ticket

https://openscience.atlassian.net/browse/ENG-896
